hypothesis-python/src/hypothesis/database.py

```diff
@@ -278,6 +417,113 @@ def delete(self, key: bytes, value: bytes) -> None:
         except OSError:
             pass
 
+    def _start_listening(self) -> None:
+        try:
+            from watchdog.events import (
+                DirCreatedEvent,
+                DirDeletedEvent,
+                DirMovedEvent,
+                FileCreatedEvent,
+                FileDeletedEvent,
+                FileMovedEvent,
+                FileSystemEventHandler,
+            )
+            from watchdog.observers import Observer
+        except ImportError:
+            warnings.warn(
+                f"listening for changes in a {self.__class__.__name__} "
+                "requires the watchdog library. To install, run "
+                "`pip install hypothesis[watchdog]`",
+                HypothesisWarning,
+                stacklevel=4,
+            )
+            return
+
+        hash_to_key = {_hash(key): key for key in self.fetch(self._metakeys_name)}
+        _metakeys_hash = self._metakeys_hash
+        _broadcast_change = self._broadcast_change
+
+        class Handler(FileSystemEventHandler):
+            def on_created(
+                _self, event: Union[FileCreatedEvent, DirCreatedEvent]
+            ) -> None:
+                # we only registered for the file creation event
+                assert not isinstance(event, DirCreatedEvent)
+                # watchdog events are only bytes if we passed a byte path to
+                # .schedule
+                assert isinstance(event.src_path, str)
+
+                value_path = Path(event.src_path)
+                # the parent dir represents the key, and its name is the key hash
+                key_hash = value_path.parent.name
+
+                if key_hash == _metakeys_hash:
+                    hash_to_key[value_path.name] = value_path.read_bytes()
+                    return
+
+                key = hash_to_key.get(key_hash)
+                if key is None:  # pragma: no cover
+                    # we didn't recognize this key. This shouldn't ever happen,
+                    # but some race condition trickery might cause this.
+                    return
```

Comment:
It's not _that_ tricky, you just need to write a new key between the load above and scheduling the observer below; if you have enough `pytest-xdist` that seems pretty likely to happen occasionally.  Well within our low standards for durability though 😅

(anyone _seriously_ distributed should be using the redis database backend, I think)